### ISSUE TITLE
RETAIN : An Interpretable Predictive Model for Healthcare using Reverse Time Attention Mechanism (2016)

### ISSUE THREAD

**Issue by JaeyeongKim00** [JaeyeongKim00/ML-DL-paper#1]
### 1. Overview

 이 논문은 Healthcare에 활용하기 용이한 machine learning model을 소개한다.

healthcare를 위한 model의 성능 지표로는 2가지가 있는데, 1. accuracy와 2. interpretability이다.

여기서 “accuracy”는 미래에 어떤 질병을 진단받을 확률을 계산하거나 예측할 때의 정확도를 뜻하며,

“interpretability”는 환자의 상태에 대한 variable들 간의 correlation을 분석할 수 있는 정도를 뜻한다.

 traditional machine learning method로 logistic regression, decision tree 같은 model을 사용할 경우, 그 과정을 분석하기 용이해 좋은 interpretability를 가지지만 accuracy가 떨어지게 된다 . 

 반면에, RNN을 사용할 경우, 훌륭한 accuracy를 보이지만, blackbox 모델이기 때문에 variable들 간의 correlation을 분석하기 어려워 poor interpretability를 가지게 된다. 

그래서 이 accuracy-interpretability trade-off를 해결하기 위해 새로운 model을 제안하는데, 이것이 RETAIN model이다. (REverse Time AttentIoN model)

 말 그대로 기존의 attention 모델과 유사하나, attention weight를 구할 때 sequential data를 reversed time order로 사용한다. (이는 의사들이 진단할 때 recent visit에 더 집중한다는 점에서 착안했다고 한다.)

따라서, 해당 논문은 RETAIN model을 통해 accuracy와 interpretability가 어떻게 향상됬는지를 보여준다.

### 2. Contribution

 기존의 attention model을 healthcare 분야에 적용할 수 있도록 time-order와 recurrence 방식을 변형해 고질적인 문제였던 accuracy-interpretability trade-off를 해결했다.

### 3.1. Introduction of variables

 training dataset으로 EHR(electronic health record)를 이용하는데, 이는 visit-order로 구성된 sequential dataset이다.

 EHR의 structure는 다음과 같다.

total N patients 중에서 $n^{th}$ patient의 $i^{th}$ visit을 ($t_{i}^{(n)}$, $x_{i}^{(n)}$)로 표현하며, 

$t_{i}^{(n)}$는 해당 patient의 time stamp, 

$x_{i}^{(n)}$는 r개의 variables(진단, 처방 등)에 대하여, 0, 1로 표현되는 r차원의 binary vector이다.

output인 $y_{i}$은 s개의 disease의 occurence에 대한 prediction으로,

$x_{i}^{(n)}$와 동일하게 0, 1로 표현되는 s차원의 binary vector이다.

### 3.2. Model structure & Comparision

  RETAIN model은 기존의 RNN attention model과 2가지 차이점을 가진다.

![Untitled](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/f92a7b19-a37a-4ee3-88e0-2b7d8cb2ef46/Untitled.png)

1. recurrence는 hidden state v vector가 아니라, attention generation component인 h와 g vector에서 일어난다.
2. reversed time order를 이용한다.

 일반 attention model의 모식도인 왼쪽 그림의 경우 recurrence가 input vector의 embedding인 $v$에서 일어나는 반면, RETAIN model의 모식도인 오른쪽 그림의 경우 recurrence가 attention generation component인 $h,g$에서 일어나게 된다. ($h,g$는 각각 RNN$_{\alpha}$와 RNN$_{\beta}$의 hidden state이다.)

RETAIN model의 모식도를 보면 다음과 같다.

![Untitled](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/cf770142-de03-4d08-9ce6-4f52b944dd19/Untitled.png)

$v, g, h, c$에 대한 수식은 다음과 같다.

![Untitled](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/83999154-2f55-421b-990c-9d6869681aa5/Untitled.png)

![Untitled](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/ef5395d3-ac32-4068-812a-794162adc0c9/Untitled.png)

![Untitled](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/3a1a9958-768c-48a0-96ae-e02656d667bd/Untitled.png)

$g$ = RNN$_{\alpha}$의 hidden layer (visit-level에서의 attention)

$h$ = RNN$_{\beta}$의 hidden layer (vatiable-level에서의 attention)

$w_{\alpha, \beta}$ = weight (training 됨)

$\alpha_{j}$ = visit-level 에서의 attention weight

$\beta_{j}$ = variance-level 에서의 attention weight

$c_{i}$ = i번째 visit에 대한 context vector (circle dot은 element wise multiplication을 뜻함.)

마지막 prediction인 $y$에 대해서는 $c$ vector에 대해 softmax 함수를 적용한다.

loss function의 경우, cross-entropy loss(multi-label classification), binary cross-entropy, MSE 등 상황에 따라 다양하게 적용될 수 있다고 한다. 해당 논문에서는 Heart failure의 여부를 classification하므로 binary cross-entorpy를 이용했다.

추가적으로, NLP에서의 standard attention의 경우에는 단어들을 RNN으로 enconding하고 attention weight를 구하는데 MLP가 쓰이는 반면, 

RETAIN model에서는 embedding 과정에서는 MLP가 사용되고 weight(attention vector)를 구하는 과정에서는 RNN이 쓰인다는 차이점이 있다고 한다.

## 3.3 Interpreting RETAIN

 RETAIN model을 해석하는데 있어서 간단하게는 위에서 다뤘듯, 

$\alpha_{j}$ = visit-level 에서의 attention weight

$\beta_{j}$ = variance-level 에서의 attention weight

이므로 각 $\alpha, \beta$ 에 따른 $y$ 변화량으로 어떤 변수가 가장 큰 contribution을 가지는지 확인할 수 있다.

 이를 수학적으로 정리하면 다음과 같다.

![Untitled](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/2c135471-dd17-4a5e-ad79-0353d1f492fb/Untitled.png)

 이 식을 정리해주면 다음과 같은 term을 얻을 수 있다.

![Untitled](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/8944067c-7ba5-4016-a7e4-7792b8f20597/Untitled.png)

 따라서, $x_{j, k}$에 어떤 coefficient를 곱해서 softmax한 값이 prediction이므로 

각 $x_{j, k}$ variable은 coefficient 만큼의 contribution을 가진다고 할 수 있다.

![Untitled](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/0d5d4560-c49a-4069-a267-d2a37d67814e/Untitled.png)

 그러므로, $x_{j, k}$의 $y_{i}$에 대한 contribution인 $\omega(y_{i}, x_{j, k})$를 다음과 같이 정의할 수 있고, contribution coefficient를 얻을 수 있어 각 variable에 대한 contribution을 분석하기 용이하다.

### 5. Experiment

 experiment는 RETAIN model과 다른 traditional model, RNN model들을 비교하여 

1. heart failure를 얼마나 잘 예측하는지, (accuracy) 
2. 결과가 각 variable의 contribution을 얼마나 잘 계산하는지 (interpretability)

를 확인한다.

![Untitled](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/5260bb1d-aa6b-47fa-9fe1-946ae6f49720/Untitled.png)

 먼저, 각 model들 간 prediction에 대한 negative log likeligood, AUC(ROC curve의 아랫부분 넓이), train/test time을 비교했다.

negative log likelihood나 AUC는 RNN based model (특히, RETAIN model)이 좋은 성능을 보인다.

 traiditional model이나 MLP의 경우 병렬 연산이 가능하기 때문에 train/test time이 상대적으로 짧은 반면, RNN based modes의 경우에는 병렬 연산이 불가능하기 때문에 train/test time 상대적으로 길다.

![Untitled](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/be6414fb-26e5-4a12-85f5-b5c8bab62700/Untitled.png)

 비교 대상이 되는 model들의 모식도이다. 

(a)는 LR, logistic regression으로 hidden layer가 없다.

(b)는 MLP로 hidden layer가 1층 있다. (256 units)

(c)는 vanilla RNN이다.

(d)는 RNN+$\alpha_M$으로, attention vector가 MLP로 만들어진다.

(e)는 RNN+$\alpha_R$로, (d)에서 attention vector가 RNN을 통해 만들어진다는 차이점이 있다.

RETAIN model은 MLP로 embedding인 $v$가 만들어지고, 2개의 attention vector(visit-level, variable-level)은 각각 RNN으로 만들어지는 구조이다.

![Untitled](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/babeb139-4051-4e5d-80bb-6e6142b4e5ed/Untitled.png)

 이 figure의 각 y axis는 각 variable들의 disease y에 대한 contribution을, x axis는 time order를 나타낸 것이다.

 (a)는 RETAIN model에 정상적으로 data sequence를 넣어준 결과이다. 실제로 HF에 큰 연관이 있는 variable(HVD, CD, CA)들은 높은 contribution을 보이며, 상대적으로 연관성이 적은 다른 variable(SD, ESL  등등. SD는 skin disease로 HF와 전혀 관련이 없다.)들은 낮은 contribution을 보인다.

그리고 실제로 HF의 probability를 약 0.25로 예측했음을 알 수 있다.

 (b)는 RETAIN model에 data seqeunce를 뒤집어서 넣어준 결과이다. 각 variable들의 contribution을 구분해내기는 했으나, (a)에 비해서 상대적으로 잘 계산해내지 못했다.

 하지만, 만약 temporality를 무시하는(time order를 무시하는) Logistic regression이나 MLP를 이용했다면 (a)와 같이 HF의 probability로 약 0.25를 예측했겠지만, 이 model의 경우 큰 contribution을 가지는 variable들의 시점이 과거로 바뀌면서 약 0.09라는 probability로 예측해냈다.

 (c)는 (a) dataset에 medication variable(AA, AC)를 추가한 형태이다. medication의 영향으로 약 0.22라는 (a)보다 낮은 HF probability를 보인 것을 알 수 있다. ...

### 6. Conclusion

 결론적으로, 기존의 attention model과 다르게 

1) attention weight를 만드는데 RNN을 적용하고 embedding에는 MLP를 적용하면서, interpretability 얻었고, 

2) 실제 의사가 진단할 때 recent visit에 더 초점을 맞춘다는 점에서 착안하여 reverse time order로 attention weight를 만들어냈고, accuracy를 향상시켰다.

 contribution 정도가 시간이 멀어짐에 따라 옅어지고, 가까울수록 커지는 상황에 대해 적용하게되면, 각 variable들의 contribution 분석이 용이하며 좋은 prediction을 만들어낼 수 있는 model로 생각할 수 있다.
